modules/alloy/src/test/ts/browser/ui/typeahead/TypeaheadModelsTest.ts

```diff
@@ -0,0 +1,362 @@
+import * as GuiSetup from 'ephox/alloy/api/testhelpers/GuiSetup';
+import { TestStore } from 'ephox/alloy/api/testhelpers/TestStore';
+import { tieredMenu as TieredMenu } from 'ephox/alloy/api/ui/TieredMenu';
+import { Typeahead } from 'ephox/alloy/api/ui/Typeahead';
+import * as MenuEvents from 'ephox/alloy/menu/util/MenuEvents';
+import * as TestDropdownMenu from 'ephox/alloy/test/dropdown/TestDropdownMenu';
+import { TestItem } from 'ephox/alloy/test/dropdown/TestDropdownMenu';
+import * as Sinks from 'ephox/alloy/test/Sinks';
+import { TypeaheadSpec } from 'ephox/alloy/ui/types/TypeaheadTypes';
+
+describe('browser.alloy.ui.typeahead.TypeaheadModelsTest', () => {
+
+  const makeTypeahead = (store: TestStore, sink: AlloyComponent, model: TypeaheadSpec['model']): SketchSpec => {
+    return Typeahead.sketch({
+      inputClasses: [ 'test-typeahead' ],
+      components: [ ],
+
+      minChars: 2,
+      model,
+      fetch: (input) => {
+        const text = Value.get(input.element).toLowerCase();
+
+        const items = Arr.range(3, (i): TestItem => {
+          return {
+            type: 'item',
+            data: {
+              value: `item-${i}-value`,
+              meta: {
+                text: `Item-${i}-text`
+              }
+            }
+          };
+        });
+
+        const future = Future.pure<TestItem[]>(items);
+
+        return future.map((f) => {
+          const items: TestItem[] = text === 'no-data' ? [
+            { type: 'separator', text: 'No data' }
+          ] : f;
+
+          const menu = TestDropdownMenu.renderMenu({
+            value: 'test-menu-value',
+            items: Arr.map(items, TestDropdownMenu.renderItem)
+          });
+          return Optional.some(TieredMenu.singleData('tiered-test-menu-name', menu));
+        });
+      },
+
+      initialData: {
+        // . for value, - for text
+        value: 'initial.value',
+        meta: {
+          text: 'initial-value'
+        }
+      },
+
+      eventOrder: Objects.wrapAll([
+        {
+          key: MenuEvents.focus(),
+          value: [ 'alloy.base.behaviour', 'tiered-menu-test' ]
+        }
+      ]),
+
+      lazySink: (c) => {
+        TestDropdownMenu.assertLazySinkArgs('input', 'test-typeahead', c);
+        return Result.value(sink);
+      },
+
+      parts: {
+        menu: TestDropdownMenu.part(store)
+      },
+
+      markers: {
+        openClass: 'typeahead-open-for-business'
+      }
+    });
+  };
+
+  const makeTypeaheadContainer = (store: TestStore, sink: AlloyComponent, model: TypeaheadSpec['model']): AlloyComponent => {
+    return GuiFactory.build({
+      dom: {
+        tag: 'div',
+        classes: [ 'overall-test-container' ]
+      },
+      components: [
+        GuiFactory.premade(sink),
+        makeTypeahead(store, sink, model)
+      ]
+    });
+  };
+
+  GuiSetup.bddAddStyles(SugarDocument.getDocument(), [
+    `.selected-item {
+      background-color: #cadbee;
+    }`
+  ]);
```

Comment:
This hook is still floating in the middle of nowhere. (though it might be a different one, I can't recall sorry)

Reply:
I checked it was the same one I commented on previously. It looks like it was moved but then the rest was dumped on top pushing it down again.

Reply:
Odd. Take two.

Reply:
Ah, I see. The original comment just mentioned moving it above something, so that's all I did. I wasn't sure if that was intentional or not. Anyway, it's at the top now.